[Wunsafe-buffer-usage] Address some false positives in handling array indices that are decidably correct

mxms0 opened the pull request:
This change aims to prevent Wunsafe-buffer-usage warnings when the code is trivially correct and we don't need to warn about it. This is done for two cases:
- When enums are used as indices 
- When the primitive array is part of a MemberExpr 

I've also adjusted the tests to match the expected behavior, which is to not warn when the code is correct. 

There's more work to be done in making this comprehensive, which I'm happy to do in a follow-up.

Commit:
[Wunsafe-buffer-usage] Fix false positives in handling enums

Do not warn if the index is an enum and we an determine statically that
it's within bounds.

rnk commented on clang/test/SemaCXX/warn-unsafe-buffer-usage-array.cpp:
```diff
@@ -39,6 +39,23 @@ void constant_idx_unsafe(unsigned idx) {
+
+void constant_enum_safe() {
+  int buffer[FooEnum::D] = { 0, 1, 2 };
+  buffer[C] = 0; // no-warning
```
A test for `buffer[A]` (0, no warn) and `buffer[D]` (warn) would be good

mxms0 replied:
Good suggestion, fixed :)

rnk commented on clang/lib/Analysis/UnsafeBufferUsage.cpp:
```diff
@@ -463,6 +463,13 @@ AST_MATCHER(ArraySubscriptExpr, isSafeArraySubscript) {
+  // Array index wasn't an integer literal, let's see if it was an enum or
+  // something similar
+  const auto IntConst = Node.getIdx()->getIntegerConstantExpr(Finder->getASTContext());
+  if (IntConst && *IntConst > 0 && *IntConst < size) {
```
Should it be `>= 0`?

mxms0 replied:
Yeeep. Good catch, should be fixed :)

Commit:
Detect when the buffer is a member access, fix tests

Add a case for when it's a member variable access and we can statically
determine the size. Also add new test to ensure the change works
reliably and update old tests to not expect this warning.

Commit:
Fix test for unsafe enum

Commit:
Minor nit, reorder expression

Commit:
Try harder to find an element we can analyze

Add logic to handle MemberExprs, so we can attempt to find the size
information for the struct member variables. This is probably still
incomplete in terms of the full space of expressions, but eh, just needs
to be better, not perfect.

rnk commented on clang/lib/Analysis/UnsafeBufferUsage.cpp:
```diff
@@ -439,8 +439,16 @@ AST_MATCHER(ArraySubscriptExpr, isSafeArraySubscript) {
       dyn_cast<StringLiteral>(Node.getBase()->IgnoreParenImpCasts());
```
There's a FIXME above about refactoring Sema::CheckArrayAccess to avoid duplication, and I'm trying to decide if that makes sense or not, or if we should just copy the logic.

It seems like the main behavior difference is that `Sema::CheckArrayAccess` doesn't go looking for a `Decl` with a `ConstantArray` type, it just looks at the type of any old expression, which means it doesn't warn on cases involving casts or other expressions with constant array type, like these:
```
int arr[1];
int oob() {return (*(int(*)[2])&arr)[1]; } // currently warns, should we keep warning or trust the cast?
int *arrayAddr()[1] { return &arr; }
int safe() { return (*arrayAddr())[0] } // currently warns, but should we warn?
```

mxms0 replied:
I think ultimately going to the Sema code is probably the better place to be, but I'm not sure who's working on that or if they have any urgency to land it. Ergo I'm in favor of landing this and working with the original owner to find the desired end state, as long as you all think that's an acceptable medium state. :)

rnk reviewed (commented):
I think there's a policy question here of how much trust we want to put into the type system. Clearly, we've already put some trust into it to reduce false positives, but we could decide to trust any old constant array type bounds, and that would be a good code simplification.

Commit:
Remove integer literal check

Simplifies the logic, as per recommendation by rnk@

mxms0 commented on clang/lib/Analysis/UnsafeBufferUsage.cpp:
```diff
@@ -463,6 +471,13 @@ AST_MATCHER(ArraySubscriptExpr, isSafeArraySubscript) {
       return true;
```
I changed the code and tests still pass, it looks like you're correct. :)

mxms0 commented:
> I think there's a policy question here of how much trust we want to put into the type system. Clearly, we've already put some trust into it to reduce false positives, but we could decide to trust any old constant array type bounds, and that would be a good code simplification.

My view is that we should trust it (the type system) until we have evidence that it's more of a problem. Developers doing really hacky bad things I expect to be a lot less common than developers doing bad things by accident (i.e. a classic index out of bounds)

mxms0 commented:
Friendly ping @rnk, @jkorous-apple :)
